### PR TITLE
Compile rustc+LLVM to wasm

### PR DESCRIPTION
**Pull request by oligamiq**
LLVM backend is now available.
It works by changing config.llvm.toml to config.toml.
This allows rustc to compile against wasm32-wasip1-threads.
https://github.com/YoWASP/clang was used as a reference.

Since wasm-ld can also be used on wasi by that project, All that remains is for browser_wasi_shim to support threading, and it should be possible to compile on the browser.

cargo is going to create a project that wraps this project like the project above, which together with rustc will generate one wasm as a driver like YoWASP/clang project.

Problem is that llvm is not patched by rust and the version of llvm is higher than rust. compiler/rustc_target/src/target_features.rs has been changed because of that.
Also, the cc library has been raised to a version that supports building to wasi.
In addition, I upgraded the wasi-sdk version.

wasm and sysroot are available at https://github.com/oligamiq/rust_wasm/
so please use them if you want to try them.

Please forgive the messy commit.

### CODE REVIEW

**Comment by bjorn3** [comment]
Thanks a lot for this PR! I'm currently at a conference, so I probably won't get to reviewing for a couple of days. If I don't review it within a week or so feel free to ping me.

File: walker.py
Comment (bjorn3) [inline walker.py:1]: Where is this file used?

File: wrapper_linker_clang.sh
Comment (bjorn3) [inline wrapper_linker_clang.sh]: I'm afraid I understand neither Chinese nor Japanese.

File: src/bootstrap/src/core/build_steps/llvm.rs
Comment (bjorn3) [inline src/bootstrap/src/core/build_steps/llvm.rs]: Isn't there a cmake toolchain file in the wasi-sdk that sets all these flags as necessary?

File: compiler/rustc_target/src/target_features.rs
Comment (bjorn3) [inline compiler/rustc_target/src/target_features.rs]: Is this because of using a different LLVM version?

File: compiler/rustc_llvm/build.rs
Comment (bjorn3) [inline compiler/rustc_llvm/build.rs]: Is it possible to avoid using LTO? It makes building a fair bit slower.

File: compiler/rustc_fs_util/src/lib.rs
Comment (bjorn3) [inline compiler/rustc_fs_util/src/lib.rs]: nit: unnecessarily removed line

File: comment.txt
Comment (bjorn3) [inline comment.txt]: Maybe use

### 
```suggestion
WASI_SDK_PATH=wasi-sdk-22.0 WASI_SYSROOT=wasi-sdk-22.0/share/wasi-sysroot WASI_CLANG_WRAPPER_LINKER="wrapper_linker_clang++.sh" ./x.py install
```

to avoid hard coding your home dir. Or if absolute paths are necessary:

```suggestion
WASI_SDK_PATH=$(pwd)/wasi-sdk-22.0 WASI_SYSROOT=$(pwd)/wasi-sdk-22.0/share/wasi-sysroot WASI_CLANG_WRAPPER_LINKER="$(pwd)/wrapper_linker_clang++.sh" ./x.py install
```

File: wrapper_linker_clang.sh
Comment (bjorn3) [inline wrapper_linker_clang.sh]: You could add `-lwasi-emulated-mman` by editing the linker args in compiler/rustc_target/src/spec/targets/wasm32_wasip1_threads.rs `-Wl,--max-memory=1073741824` is also produced there.

File: wrapper_linker_clang.sh
Comment (bjorn3) [inline wrapper_linker_clang.sh]: Should this just remove the `-Wl,` prefix? I was genuinely running out of memory without this argument. Wasm shared memories need to declare a maximum size and lld gives a very small size by default.

**Review by bjorn3 — CHANGES_REQUESTED** [review]
Sorry for the late reply. I've rebased the wasm changes and created the new compile_rustc_for_wasm16 branch for this. Would you mind rebasing and changing the target of this PR to that branch? You can use the same edit button that allows changing the PR title and then clicking on the button next to "oligamiq wants to merge 86 commits into". Also squashing your commits into either a single commit or a logical division of commits would be appreciated.

File: walker.py
Comment (oligamiq) [reply walker.py:1]: It searches for symbols to identify .c files and other files that use symbols that cannot be used in WASM. It was useful for debugging, but can be eliminated.

**Comment by oligamiq** [comment]
Since I was assuming version 15, there was a conflict. 
rustc is quite lax in that area, so I'll modify it to directly add -lwasi-emulated-mman.

**Comment by bjorn3** [comment]
You can drop the first three commits that are currently in this PR when rebasing on top of compile_rustc_for_wasm16.

**Comment by oligamiq** [comment]
I've shut down the work server, so I'll do it tomorrow. I'll give it a try.

**Comment by bjorn3** [comment]
Looks like the first commit is still there. If you want I could do the rebase for you and if you enable the github option for this have me push it directly to your PR branch.

**Comment by oligamiq** [comment]
What is the 'first commit'?

**Comment by bjorn3** [comment]
I'm referring to https://github.com/bjorn3/rust/pull/6/commits/295aca58b83ab0ca0fd766fd22aafc80933dfc95. This commit originates from the compile_rustc_for_wasm15 branch of mine despite this PR now targetting the compile_rustc_for_wasm16 branch.

**Comment by oligamiq** [comment]
If we lose it, won't it cause the build to possibly fail or fail to compile on the cranelift backend?
Or are you talking about integrating commits?

**Comment by oligamiq** [comment]
Sorry. It makes me crazy because it already exists. I'll turn it off.

Add Message
I'm currently adding `-lwasi-emulated-mman` and trying to build from scratch, so I can't rebase right now. If it succeeds, I'll do it with the new commit. `wrapper_linker_clang.sh` should be gone.

**Comment by oligamiq** [comment]
I believe the work is complete.

File: compiler/rustc_target/src/spec/targets/wasm32_wasip1_threads.rs
Comment (bjorn3) [inline compiler/rustc_target/src/spec/targets/wasm32_wasip1_threads.rs]: For the Cc::No case the `-Wl,` prefix has to be omitted.

```suggestion
        &["--import-memory", "--export-memory", "--shared-memory", "--max-memory=1073741824", "-lwasi-emulated-mman"],
```

File: compiler/rustc_target/src/target_features.rs
Comment (bjorn3) [reply compiler/rustc_target/src/target_features.rs]: Seems so.

**Comment by bjorn3** [comment]
It seems like it still isn't rebased on top of the compile_rustc_for_wasm16 branch. I've got it rebased locally and am currently building it to try it out. If you enable the option "Allow edits and access to secrets by maintainers" as explained by https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork I can push my rebased version to this PR.

**Comment by oligamiq** [comment]
ok, understand. You want to create a new branch from wasm16 and cherry-pick it. I've given you the permissions for now, so please take care of it this time.

I thought github would do it automatically...

**Comment by bjorn3** [comment]
Pushed

> I thought github would do it automatically...

I think github preserves your choice, so if you ever disabled it, it will be kept disabled for all future PR's by default.

**Comment by oligamiq** [comment]
No, I thought I could just apply the changes to the commit in a nice way regardless of the top of branch.

**Comment by bjorn3** [comment]
There were some conflicts between the top of the branch and your commits, so it wouldn't have been possible to merge until those conflicts were resolved. I resolved them for you while rebasing.

**Comment by bjorn3** [comment]
I'm getting the following locally when building LLVM:

```
/home/bjorn/Projects/rust/src/llvm-project/llvm/include/llvm/ADT/bit.h:52:10: fatal error: 'machine/endian.h' file not found
   52 | #include <machine/endian.h>
      |          ^~~~~~~~~~~~~~~~~~
```

Edit: Messed up the llvm-project submodule while I did the rebase.

**Comment by oligamiq** [comment]
I believe there was a .define("HAVE_DLOPEN", ""); followed by a .define to turn it OFF.

I'll go look for it.

I have restored it, though,
https://github.com/oligamiq/rust/blob/compile_rustc_for_wasm15_dev_keep/src/bootstrap/src/core/build_steps/llvm.rs
Am I misremembering?

I understand, as I also had trouble with submodule changes.

**Comment by oligamiq** [comment]
about [Fix linking stdc++ on wasi]

config.llvm.toml:
It should work by simply setting up the wasi-sysroot, but it may not have worked correctly due to using version 24.

src/bootstrap/src/core/build_steps/compile.rs:
By upgrading the version of `cc`, linking for the WASI target should also function properly, and by setting the WASI_SYSROOT environment variable, linking to `libc++.a` should automatically occur. However, this might also be influenced by the use of version 24.
https://github.com/rust-lang/cc-rs/pull/1114

**Comment by bjorn3** [comment]
In this case the `LLVM_STATIC_STDCPP` env var ended up with just `libstdc++.a` instead of the full path, causing the rustc_llvm build script to produce an empty search path to find `libstdc++.a` in, which then results in rustc giving an error about the empty search path argument. I already have WASI_SYSROOT set.

**Comment by bjorn3** [comment]
I finally managed to get it to work locally. It does deadlock most of the time when running in wasmtime however unless I use `RUST_LOG=trace`. Maybe setting this slows down something just enough to avoid a deadlock even though it doesn't cause any additional output from wasmtime?

**Comment by oligamiq** [comment]
really?
This is what I copied during this pull request:
https://github.com/bjorn3/rust/pull/7

It should have worked without any issues, but I will also test it with the copied version to ensure it works.
https://github.com/oligamiq/rust_wasm/tree/main/rustc_llvm

Add:
copy on rust_wasm is normal operation.
I will clone the repository, compile it, and check if it works correctly on this branch.

**Comment by bjorn3** [comment]
Updating Wasmtime seems to have fixed the deadlocks.

**Comment by oligamiq** [comment]
oh...

**Comment by bjorn3** [comment]
Checked that this PR doesn't break compiling cg_clif for single threaded wasm32-wasip1. I've spent enough time on this for today. I'm going to take another look at the changes tomorrow to see if anything is unnecessary or can be improved. After that I will merge this PR. Thanks again for working on this!

**Comment by bjorn3** [comment]
This should be updated to use the https://github.com/YoWASP/llvm-project/releases/tag/rustc-2024-09-24-do-not-remove tag for the llvm-project submodule to ensure that it remains possible to build in the future.

**Comment by oligamiq** [comment]
I think LLVM_ENABLE_THREADS=OFF was broken

**Comment by bjorn3** [comment]
Indeed. That commit was just a couple of changes that are necessary anyway to compile for wasm32-wasip1 once LLVM_ENABLE_THREADS=OFF gets fixed.

**Comment by oligamiq** [comment]
Also, RustCompiler have code to check if LLVM has multithreaded.

https://llvm.org/doxygen/group__LLVMCCoreThreading.html

**Pull request merged**
